### PR TITLE
Use nvidia gpu operator for nvidia-gpu flavor

### LINKED ISSUE #1017: Use NVIDIA's gpu-operator for GPU node support
/kind feature

**Describe the solution you'd like**

#1002 implemented the "nvidia-gpu" flavor via `postKubeadmCommands` recommended by NVIDIA, as explained in [this comment](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/426#issuecomment-626966868). 

But NVIDIA's [gpu-operator](https://github.com/NVIDIA/gpu-operator) seems like a cleaner, more future-proof solution. We should investigate whether it supports `containerd` now and whether the current implementation could be replaced with gpu-operator.

**Anything else you would like to add:**

See the discussion in #426 and the current implementation in #1002.

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by shysank**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This pr updates `nvidia-gpu` flavor to use gpu-operator for installing and managing gpu components. `gpu-operator` is installed using cluster resource set.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1017 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Use nvidia gpu operator for nvidia-gpu flavor to manage gpu components
```

### CODE REVIEW

**Comment by shysank** [comment]
Ran e2e tests with gpu enabled, and they are passing => https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/1254/pull-cluster-api-provider-azure-e2e/1376554297197596672

CRS created as part of the flavor now:

```
INFO: "with a single control plane node and 1 node" started at Mon, 29 Mar 2021 15:26:13 UTC on Ginkgo node 1 of 3
[1mSTEP[0m: Creating a namespace for hosting the "create-workload-cluster" test spec
INFO: Creating namespace create-workload-cluster-uou803
INFO: Creating event watcher for namespace "create-workload-cluster-uou803"
INFO: Creating the workload cluster with name "capz-e2e-ekv16o" using the "nvidia-gpu" template (Kubernetes v1.19.7, 1 control-plane machines, 1 worker machines)
INFO: Getting the cluster template yaml
INFO: clusterctl config cluster capz-e2e-ekv16o --infrastructure (default) --kubernetes-version v1.19.7 --control-plane-machine-count 1 --worker-machine-count 1 --flavor nvidia-gpu
INFO: Applying the cluster template yaml to the cluster
cluster.cluster.x-k8s.io/capz-e2e-ekv16o serverside-applied
azurecluster.infrastructure.cluster.x-k8s.io/capz-e2e-ekv16o serverside-applied
kubeadmcontrolplane.controlplane.cluster.x-k8s.io/capz-e2e-ekv16o-control-plane serverside-applied
azuremachinetemplate.infrastructure.cluster.x-k8s.io/capz-e2e-ekv16o-control-plane serverside-applied
machinedeployment.cluster.x-k8s.io/capz-e2e-ekv16o-md-0 serverside-applied
azuremachinetemplate.infrastructure.cluster.x-k8s.io/capz-e2e-ekv16o-md-0 serverside-applied
kubeadmconfigtemplate.bootstrap.cluster.x-k8s.io/capz-e2e-ekv16o-md-0 serverside-applied
clusterresourceset.addons.cluster.x-k8s.io/crs-gpu-operator serverside-applied
configmap/nvidia-clusterpolicy-crd serverside-applied
configmap/nvidia-gpu-operator-components serverside-applied
configmap/cni-capz-e2e-ekv16o-crs-0 serverside-applied
clusterresourceset.addons.cluster.x-k8s.io/capz-e2e-ekv16o-crs-0 serverside-applied
```

**Comment by shysank** [comment]
This pr is ready to be reviewed now.

File: test/e2e/azure_test.go [hunk 1]
```diff
@@ -41,7 +41,7 @@ var _ = Describe("Workload cluster creation", func() {
 		specName      = "create-workload-cluster"
 		namespace     *corev1.Namespace
 		cancelWatches context.CancelFunc
-		result       *clusterctl.ApplyClusterTemplateAndWaitResult
+		result        *clusterctl.ApplyClusterTemplateAndWaitResult
```
Comment (CecileRobertMichon) [inline test/e2e/azure_test.go:44]: nit: extra space
```suggestion
		result       *clusterctl.ApplyClusterTemplateAndWaitResult
```

File: test/e2e/azure_test.go [hunk 2]
```diff
@@ -323,6 +323,7 @@ var _ = Describe("Workload cluster creation", func() {
 				WaitForClusterIntervals:      e2eConfig.GetIntervals(specName, "wait-cluster"),
 				WaitForControlPlaneIntervals: e2eConfig.GetIntervals(specName, "wait-control-plane"),
 				WaitForMachinePools:          e2eConfig.GetIntervals(specName, "wait-machine-pool-nodes"),
+				Args:                         []string{"--server-side"},
```
Comment (CecileRobertMichon) [inline test/e2e/azure_test.go:326]: could you add a code comment to explain why this is needed

**Review by mboersma — COMMENTED** [review]
Docs changes look great. I've skimmed the YAML and Go changes and don't see anything other than what @CecileRobertMichon already pointed out.

File: test/e2e/azure_test.go (hunk 1 above)
Comment (shysank) [reply test/e2e/azure_test.go:44]: Noticed that this space gets added when I format the code in goland. I think it's okay to leave it. wdyt?

File: test/e2e/azure_test.go [hunk 3]
```diff
@@ -323,6 +323,11 @@ var _ = Describe("Workload cluster creation", func() {
 				WaitForClusterIntervals:      e2eConfig.GetIntervals(specName, "wait-cluster"),
 				WaitForControlPlaneIntervals: e2eConfig.GetIntervals(specName, "wait-control-plane"),
 				WaitForMachinePools:          e2eConfig.GetIntervals(specName, "wait-machine-pool-nodes"),
+				// nvidia-gpu flavor creates a config map as part of a crs, that exceeds the annotations size limit when we do kubectl apply.
```
Comment (CecileRobertMichon) [inline test/e2e/azure_test.go:326]: how come we're not running into this when using the flavor outside of e2e? Is there a chance users will run into this if using `clusterctl` with --flavor gpu?

Comment (shysank) [reply test/e2e/azure_test.go:326]: That's a great point. I think it should be updated in the docs. Will do it now.

**Comment by shysank** [comment]
@CecileRobertMichon I've addressed all the comments, PTAL a look whenever you get a chance.

**Comment by nader-ziada** [comment]
To make sure the GPU test that has been failing in the peridic job works here: 

/test pull-cluster-api-provider-azure-e2e-full

**Pull request merged**